File: Algorithm_237_Delete Node in a Linked List.cpp
#include <iostream>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
void deleteNode(ListNode* node)
{
	ListNode *tmp=node,*parent;
	if (!tmp) return;
	while (tmp->next)
	{
		tmp->val=tmp->next->val;
		parent=tmp;
		tmp=tmp->next;
	}
	parent->next=NULL;
	delete tmp;
}
int main()
{
	ListNode *n1=new ListNode(1);
	ListNode *n2=new ListNode(2);
	ListNode *n3=new ListNode(3);
	ListNode *n4=new ListNode(4);
	n1->next=n2;
	n2->next=n3;
	n3->next=n4;
	deleteNode(n2);
	ListNode *tmp=n1;
	while (tmp)
	{
		cout<<tmp->val<<' ';
		tmp=tmp->next;
	}
}

File: Algorithm_237_Delete Node in a Linked List.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

// 思路：由于题干保证给定的node不是链表的最后一个节点，因此直接把该节点的val置为其下一个节点的val并修改next使其指向再下一个节点

class Solution {
    public void deleteNode(ListNode node) {
        node.val = node.next.val;
        node.next = node.next.next;
    }
}